File: go/mission/belief.go
package mission

// State estimation — observations are not facts. The CRM, the inbox, a ticket and a human disagree;
// the runtime fuses possibly-conflicting observations of one key into a confidence-weighted BELIEF,
// flagging conflict when high-confidence sources diverge. Go port of agentic_os/mission/belief.py.

import (
	"math"
	"sort"
)

// Fuse fuses observations [{value, source, confidence}] of ONE key into a single Belief. Group by
// value, sum source-confidence per candidate, pick the strongest. confidence = agreement (winner
// mass / total) × reliability (best single source, capped at 1) so BOTH a genuine disagreement and
// a single uncertain source read as low-confidence. conflict when a competing value carries >= 0.34
// of total mass.
func Fuse(observations []map[string]any) Belief {
	if len(observations) == 0 {
		return Belief{Value: nil, Confidence: 0.0}
	}
	mass := map[string]float64{}
	maxconf := map[string]float64{}
	src := map[string][]string{}
	latest := map[string]float64{}
	valueOf := map[string]any{}
	var order []string // first-seen order, for stable tie-breaking (matches Python dict order)

	for _, o := range observations {
		key := hashKey(o["value"])
		if _, seen := valueOf[key]; !seen {
			order = append(order, key)
		}
		valueOf[key] = o["value"]
		c := toFloat(o["confidence"], 1.0)
		mass[key] += c
		if c > maxconf[key] {
			maxconf[key] = c
		}
		s, _ := o["source"].(string)
		if s == "" {
			s = "?"
		}
		if !strContains(src[key], s) {
			src[key] = append(src[key], s)
		}
		if ts := toFloat(o["ts"], 0.0); ts > latest[key] {
			latest[key] = ts
		}
	}

	total := 0.0
	for _, m := range mass {
		total += m
	}
	if total == 0 {
		total = 1.0
	}

	keys := append([]string{}, order...)
	sort.SliceStable(keys, func(i, j int) bool {
		if mass[keys[i]] != mass[keys[j]] {
			return mass[keys[i]] > mass[keys[j]]
		}
		return latest[keys[i]] > latest[keys[j]]
	})

	winKey := keys[0]
	runnerMass := 0.0
	if len(keys) > 1 {
		runnerMass = mass[keys[1]]
	}
	agreement := mass[winKey] / total
	reliability := math.Min(1.0, maxconf[winKey])
	return Belief{
		Value:      valueOf[winKey],
		Confidence: roundN(agreement*reliability, 3),
		Sources:    src[winKey],
		Conflict:   runnerMass >= 0.34*total,
		UpdatedAt:  now(),
	}
}

File: agentic_os/mission/belief.py
"""State estimation — observations are not facts.

The CRM, the inbox, a support ticket and a human will disagree. The runtime never writes an
observation straight into world state; it fuses possibly-conflicting observations of one key
into a confidence-weighted BELIEF, flagging conflict when high-confidence sources diverge. A
capability reads the belief; the planner can insert a disambiguation node when confidence is
low or `conflict` is set. This is the robotics sensor-fusion pattern for enterprise data.
"""
from __future__ import annotations

from collections import defaultdict
from typing import Any

from .types import Belief, DecisionEvidence, now


def fuse(observations: list[dict], key: str = "") -> Belief:
    """Fuse observations [{value, source, confidence, source_type?, source_ref?}] of ONE key
    into a single Belief.

    Strategy: group by value, sum source-confidence per candidate, pick the strongest.
    Confidence = winner_mass / total_mass. `conflict` when a competing value also carries
    substantial (>=0.34 of total) mass — i.e. the sources genuinely disagree. Selection is by
    confidence mass only: NO ``source_type`` is privileged (never 'regex wins' / 'model wins').
    Each observation is also captured as a ``DecisionEvidence`` on the belief.
    """
    if not observations:
        return Belief(value=None, confidence=0.0, sources=[])

    evidence = [
        DecisionEvidence(
            field=key or o.get("key", ""),
            value=o.get("value"),
            source_type=o.get("source_type", "prior"),
            source_ref=o.get("source_ref", ""),
            confidence=float(o.get("confidence", 1.0)),
        )
        for o in observations
    ]

    mass: dict[Any, float] = defaultdict(float)
    maxconf: dict[Any, float] = defaultdict(float)   # reliability: the best single source per value
    src: dict[Any, list[str]] = defaultdict(list)
    latest: dict[Any, float] = defaultdict(float)
    value_of: dict[Any, Any] = {}
    for o in observations:
        v = o.get("value")
        key = _hashable(v)
        value_of[key] = v
        c = float(o.get("confidence", 1.0))
        mass[key] += c
        maxconf[key] = max(maxconf[key], c)
        s = o.get("source", "?")
        if s not in src[key]:
            src[key].append(s)
        latest[key] = max(latest[key], o.get("ts", 0.0) or 0.0)

    total = sum(mass.values()) or 1.0
    ranked = sorted(mass.items(), key=lambda kv: (kv[1], latest[kv[0]]), reverse=True)
    win_key, win_mass = ranked[0]
    runner_mass = ranked[1][1] if len(ranked) > 1 else 0.0
    # confidence = agreement across sources × reliability of the best source (capped at 1.0),
    # so BOTH a genuine disagreement and a single uncertain source read as low-confidence.
    agreement = win_mass / total
    reliability = min(1.0, maxconf[win_key])
    return Belief(
        value=value_of[win_key],
        confidence=round(agreement * reliability, 3),
        sources=src[win_key],
        conflict=runner_mass >= 0.34 * total,   # a competing value carries substantial mass
        updated_at=now(),
        evidence=evidence,
    )


def _hashable(v: Any):
    try:
        hash(v)
        return v
    except TypeError:
        return repr(v)
